### PR TITLE
Portability improvements

### PR DESCRIPTION
**Pull request by dkochmanski**
With these commits cl-quil loads without problems on ECL[1]. Some fixes are made to improve CCL support, but it still doesn't load cl-quil because of how ADT creates structuers (it uses uninterned symbols and make-load-form-saving-slots on CCL doesn't use slot offsets but names).

[1] Branch is not merged yet, but it introduces improvements to make-load-form-saving-slots and adds custom equivalence predicate capabilities to make-hash-table.

https://gitlab.com/embeddable-common-lisp/ecl/merge_requests/152

### CODE REVIEW

**Comment by dkochmanski** [comment]
how complete is the test suite? when running tests against ecl I have some weird failures, most importantly:

```
Condition of type: SIMPLE-ERROR
No applicable method for CL-QUIL::PRINT-INSTRUCTION-GENERIC with arguments of types 
 LONG-FLOAT
 STRING-STREAM.
```

And indeed, there is no such specialization. When this error is hit (after many successfully passed tests), selecting restart `1. (TRY-NEXT-COMPILER) Ignore this error and try the next compiler in the list.` gives me the same error over and over again.

There is also this simple error earlier:

```
Condition of type: SIMPLE-ERROR
No keys match in SWITCH. Testing against (NIL :FLUSHING) with EQUAL.
```

which stems from `compress-instructions` -- `(old-state (governed-queue-state governed-queue))` old state is NIL.

**Comment by dkochmanski** [comment]
cl-quil test results (for the reference)

[xxx.txt](https://github.com/rigetti/quilc/files/3212862/xxx.txt)

**Comment by dkochmanski** [comment]
for quilc compilation I've made a PR to https://github.com/rigetti/rpcq/pull/69 which drops dependency on unicly (for uuid which is much better written and more portable).

**Comment by ecpeterson** [comment]
You're probably right that quilc unwisely mixes `double-float` and `long-float`, leading to some of the problems here. However, I don't see where `print-instruction-generic` might be being called on an unboxed float of any sort. Any chance of sharing a call graph?

The `(NIL :FLUSHING)` error is more concerning. I'll think about that and get back to you.

File: benchmarking/rewiring-analysis.lisp [hunk 1]
```diff
@@ -134,13 +134,13 @@
             (read-quil-file prog-source))))
 
        (by-assignment (prog-source chip assn)
-         #+sbcl (sb-ext:gc :full t)
+         (trivial-garbage:gc :full t)
```
Comment (stylewarning) [inline benchmarking/rewiring-analysis.lisp]: @ecpeterson I wonder if this was a hold-over from Corwin? I wonder if we can nix this...

File: src/analysis/rewrite-arithmetic.lisp [hunk 2]
```diff
@@ -119,7 +119,8 @@ expressions, as elements of a single new memory descriptor.
          ;; just a single name (i.e., more than just __P). It adds
          ;; little bit of complexity in favor of a bit of additional
          ;; extensibility.
-         (recalculation-table (make-hash-table :test 'memory-ref=))
+         (recalculation-table (make-hash-table :test 'memory-ref=
+                                               :hash-function 'memory-ref-hash))
```
Comment (stylewarning) [inline src/analysis/rewrite-arithmetic.lisp]: is ECL ok with this extension?

File: src/package.lisp [hunk 3]
```diff
@@ -422,7 +422,10 @@
   (:export
    #:standard-qubit-relabeler           ; FUNCTION
    )
-  )
+  #+ (or ccl ecl sbcl)
+  (:import-from
+   #+ccl #:ccl #+ecl #:ext #+sbcl #:sb-ext
+   double-float-positive-infinity))
```
Comment (stylewarning) [inline src/package.lisp]: Could we instead do a `defconstant` in the utilities file? I want to error at compile time if this isn't defined.

File: src/parser.lisp [hunk 4]
```diff
@@ -95,9 +95,9 @@
    (return (tok :PARAMETER (param $1))))
   ((eager "{{STRING}}")
    (return (tok :STRING (read-from-string $@))))
-  ((eager #.(string #\DAGGER))
+  ((eager #.(string #-ccl #\DAGGER #+ccl #\u+2020))
```
Comment (stylewarning) [inline src/parser.lisp]: haha :(

File: src/analysis/rewrite-arithmetic.lisp (hunk 2 above)
Comment (dkochmanski) [reply src/analysis/rewrite-arithmetic.lisp]: I've implemented it yesterday (see the linked branch). hash-table with custom test is also used in god-table.

File: src/parser.lisp (hunk 4 above)
Comment (dkochmanski) [reply src/parser.lisp]: I don't understand why this is funny (in a sad fashion ;), could you elaborate?

File: src/package.lisp (hunk 3 above)
Comment (dkochmanski) [reply src/package.lisp]: fixed (I've used alexandria's defie-constant), if not implemented the value resolves to (error …).

**Comment by dkochmanski** [comment]
@ecpeterson the problem with `(NIL :FLUSHING)` was caused by calling make-instance on a structure-class designator. Common Lisp is silent about the results of such operation. Last commit addresses this issue and `TEST-COMPILING-EMPTY-PROGRAM` passes now.

I'm going to look now into printing issue.

**Comment by dkochmanski** [comment]
So the problem with printing is caused by the assumption that PI is of type double-float. `build-gate` takes the number and produces long float instead of a boxed value in this case. Shadowing pi with a constant #.(coerce cl:pi 'double-float) solves the issue (pushed). Now I'm facing a regression (in ECL) with weak hash tables.

**Comment by dkochmanski** [comment]
With current state of this branch (most notably pi shadowed to be always double-float) ECL as compiled from branch `relax-reqs` passes all tests (except the test where c stack is exceeded and can't be grown any further). Alternative solution to shadowing pi would not be using this variable (but i.e a constant `+pi-df+`) or making quilc work with long floats too.

ECL performance is poor. Completing all tests took ~2h (compared to SBCL's ~15m). Everything was compiled on default optimization settings so it could be probably way better (I need to think about re-tweaking default compiler policies). Which operation takes most time depends on a test (sometimes it is gc, sometimes equal/string hashtable access, generic function dispatch, aref etc).

I'm attaching file with the test log. I don't know why fiasco nicely aligns test output on sbcl and not on ECL (format issue? using implementation extension on sbcl?).

[results-ecl.txt](https://github.com/rigetti/quilc/files/3217507/results-ecl.txt)
[results-sbcl.txt](https://github.com/rigetti/quilc/files/3217508/results-sbcl.txt)

File: benchmarking/rewiring-analysis.lisp (hunk 1 above)
Comment (ecpeterson) [reply benchmarking/rewiring-analysis.lisp]: This line indeed seems unimportant. You'd want it if you were measuring wall time inside of this loop, but we aren't.

File: src/compressor/compressor.lisp [hunk 5]
```diff
@@ -752,7 +752,7 @@ This specific routine is the start of a giant dispatch mechanism. Its role is to
   (let* ((output nil)
          (n-qubits (chip-spec-n-qubits chip-specification))
          (governors (make-list (length (chip-specification-objects chip-specification))))
-         (global-governor (make-instance 'governed-queue))
+         (global-governor (make-governed-queue))
```
Comment (ecpeterson) [inline src/compressor/compressor.lisp]: 👍

File: src/parser.lisp (hunk 4 above)
Comment (ecpeterson) [reply src/parser.lisp]: I can't explain why it's funny, but the humor is in how several thousands of LoC doing the heavy lifting of quantum compilation mostly all just worked†, but this goofy easter egg that no one actually uses required special treatment.

† - "Just worked" only after a month of very serious work at getting MAGICL into line.

Comment (notmgsk) [reply src/parser.lisp]: Wait - does this mean that the literal "DAGGER" won't work on CCL?

Comment (ecpeterson) [reply src/parser.lisp]: CCL doesn't know the human-readable mnemonic `#\DAGGER` for 🗡, which we make available as an alias for the Quil-literal DAGGER.

Comment (notmgsk) [reply src/parser.lisp]: Oh, the string "DAGGER" is caught elsewhere. My bad.

**Comment by appleby** [comment]
@dkochmanski Regarding the stack overflow issue, does ECL do tail-call-optimization with default optimization settings?

This is speculation, but I have seen stack overflow in OUTER-INSTRUCTION-LOOP in the file `compressor.lisp` when running tests compiled with `(debug 3)` on SBCL, presumably because that disables TCO on SBCL.

File: benchmarking/rewiring-analysis.lisp (hunk 1 above)
Comment (dkochmanski) [reply benchmarking/rewiring-analysis.lisp]: in this case should it stay or should it go?

File: src/parser.lisp (hunk 4 above)
Comment (dkochmanski) [reply src/parser.lisp]: this is a minor issue (strings with dagger will still work whatsoever), the biggest (and as far as I'm aware the only) blocker for CCL is the fact that you do #.(foo) on structures defined with `adt:defdata`. This macro creates structures with uninterned slot names and CCL can't load them back (ECL couldn't neither but I've fixed it by using offsets in `make-load-data-using-slots`).

**Comment by dkochmanski** [comment]
@appleby whenever ECL does TCO or not depends on the underlying C compiler. On GCC and Clang with speed >=2 flag "-O2" is set and optimization is present. That said I've build ECL with debugging enabled for this test run, so I'll rebuild all with more favourable optimize settings.

File: src/parser.lisp (hunk 4 above)
Comment (stylewarning) [reply src/parser.lisp]: @dkochmanski I made a PR to CL-ADT to fix that issue

File: benchmarking/rewiring-analysis.lisp (hunk 1 above)
Comment (stylewarning) [reply benchmarking/rewiring-analysis.lisp]: let's nix it

**Comment by stylewarning** [comment]
Could we also state in the README the version needed for ECL and CCL?

**Comment by dkochmanski** [comment]
I've removed call to gc.

regarding the README, Quilc does not compile on CCL (due to a reason I've mentioned with structure load-forms and uninterned slot names) and with ECL changes necessary for it to work are still in a branch. Even after merge to develop they won't be released yet, because develop branch is mid-release branch (so 16.1.3 is not enough to compile quilc and 16.2.0 is not released yet).

**Comment by dkochmanski** [comment]
compilling with optimization options (speed 3) (debug 0) (safety 1) does not help with the stack overflow. problem happens in test-compiler-hook test (not right of the bat, some test cases pass until the one which consumes too much of the stack, resizing helps and few other pass until we hit a wall).

**Pull request merged**